### PR TITLE
Implement pass by reference for large objects

### PR DESCRIPTION
**Pull request by boothinator**
Detect when the item type T is large and use pass-by-reference instead of pass-by-value for push() and unshift(). This should be faster for large objects, per the [C++ Core Guidelines F.16](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const). Additionally, provide implementations of pop(), shift(), first(), and last() that take a reference parameter for these large types.

This feature is implemented similarly to the IT parameter. The new template parameter TIn stores the input type for push() and unshift(). The TIn parameter defaults to the newly created Helper::Input::Type, which is `T` when `sizeof(T) <= 2*sizeof(void *)`, and is `const T &` when `sizeof(T) > 2*sizeof(void *)`. The other access methods are unchanged, although new versions are provided that take a reference to avoid having to return large objects, if desired by the caller.

### CODE REVIEW

**Comment by rlogiacco** [comment]
Thanks for the contribution. Would you please also add an Example and a section in the README describing when and how this  should/can be used or which benefits/drawbacks it introduces?

File: CircularBuffer.h [hunk 1]
```diff
@@ -102,27 +119,55 @@ template<typename T, size_t S, typename IT = typename Helper::Index<(S <= UINT8_
 	 */
 	T shift();
 
+	/**
+	 * @brief Removes an element from the beginning of the buffer.
+	 *
+	 * @warning Calling this operation on an empty buffer has an unpredictable behaviour.
+	 */
+	void shift(T &);
```
Comment (rlogiacco) [inline CircularBuffer.h:127]: how is a user protected against using the wrong version of the shift function if they are both available?

**Comment by boothinator** [comment]
Thanks for the feedback. I'll make some changes to address your comments.

**Comment by boothinator** [comment]
How are you determining that "580 bytes" number? Is there a specific piece of code that you're using to produce that number?

Also, I added a benchmark example where I'm directly comparing the performance of pass by value to pass by reference. One thing I noticed is that pass by reference is faster in every circumstance, even for something small like uint8_t. Would you consider making every argument a reference, regardless of item size?

That being said, I do want to go through the disassembly a bit to understand what GCC is producing better. I was expecting pass by value to be faster than pass by reference for small data types. Perhaps it's a quirk of the AVR architecture, or GCC, or the default optimization settings, or all of them.

File: CircularBuffer.h (hunk 1 above)
Comment (boothinator) [reply CircularBuffer.h:127]: There is no wrong version of shift(), just user preference.

**Comment by rlogiacco** [comment]
Hey @boothinator , are you still willing to merge this PR in the main branch?

>  Would you consider making every argument a reference, regardless of item size?

Sure, can you foresee any undesirable effect for those using a previous version of the library?

**Comment by boothinator** [comment]
It's probably safest to leave things the way they are.

**Pull request merged**